feat(stories): enable rich text in story editor

garcia5 opened the pull request:
- add [TipTap](https://tiptap.dev/docs/editor/getting-started/install/react) dependency (and some extensions) to frontend
- replace `TextArea` component with custom `TipTapEditor` component for story editing
    - supports inline markdown + automatic links
    - links are created with `rel="noopener noreferrer nofollow" target="_blank"` attributes by default, see [Link extension docs](https://tiptap.dev/docs/editor/extensions/marks/link)
    - saves stories as html text
- use `TipTapEditor` with editing _disabled_ to render story html in `Story` component

jboolean reviewed (changes requested):
Thanks for trying this out! I'm seeing some immediate problems on the preview branch

1. The overflow behavior of the editor changed

Here it is before:

https://github.com/user-attachments/assets/70fc90fd-c0a3-4957-b581-3ee975223df9
And now:
https://github.com/user-attachments/assets/577bba55-9703-4a2a-8d8a-036126e37887

I tried to render an existing story with a url in it to see if the url worked (photo: `nynyma_rec0040_1_00998_0026`).
Rendering before:
<img width="831" height="394" alt="image" src="https://github.com/user-attachments/assets/7ce93c8f-ba0d-4f38-b636-dc1b43b97cb9" />

Rendering now:
<img width="847" height="284" alt="image" src="https://github.com/user-attachments/assets/c09406f3-71c7-4c68-b96b-1fd3b2609643" />

The url is not clickable and the line breaks are missing. 

I noticed that it persists in html rather than markdown. I think Markdown is a better choice here
- It is a much simpler format. We only needs links and basic formatting. Html is overly complex for our needs.
- Markdown is a simple, human readable format. Persisted content where the user has not added any formatting is cluttered with p tags.
- Markdown is backwards compatible with plaintext. Meaning, the plaintext content we already have can be rendered as markdown. Especially if users used links, they can be rendered, as can basic formatting like underscores and asterisks. 
- Not a trap door. A markdown renderer can be removed and the content still looks good. HTML would require a migration.
- Security. I'm sure tiptap renderer has xss mitigations, but it still makes me nervous about what happens if a story is submitted (especially bypassing the editor) with script tags, images, etc. I feel more confident in React's built-in escapes that are currently used on the plaintext content. fwiw I did try to put it script and img tags and the tiptap renderer stripped them out, and the header rendered as a p.

I see tiptap does support markdown to some extent.
I think it's also worth looking at `react-markdown` for rendering. If you enable the gfm (Github flavored markdown) plugin in react-markdown it appears to enable auto-linking. It also support allow-listing tags.

garcia5 commented:
> Thanks for trying this out! I'm seeing some immediate problems on the preview branch
> 
> 1. The overflow behavior of the editor changed
> 
> Here it is before:
> 
> https://github.com/user-attachments/assets/70fc90fd-c0a3-4957-b581-3ee975223df9 And now: https://github.com/user-attachments/assets/577bba55-9703-4a2a-8d8a-036126e37887
> 
> I tried to render an existing story with a url in it to see if the url worked (photo: `nynyma_rec0040_1_00998_0026`). Rendering before: <img alt="image" width="831" height="394" src="https://private-user-images.githubusercontent.com/406931/528988414-7ce93c8f-ba0d-4f38-b636-dc1b43b97cb9.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc1NTU1MTYsIm5iZiI6MTc2NzU1NTIxNiwicGF0aCI6Ii80MDY5MzEvNTI4OTg4NDE0LTdjZTkzYzhmLWJhMGQtNGYzOC1iNjM2LWRjMWI0M2I5N2NiOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMTA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDEwNFQxOTMzMzZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jNmFkMjMwMDNkYTY2ZjE2NjM1ODdmOWZmMGY3YjMzNjFlMjg0YTg4OTc2Y2UzOTViNTEwZDA2YzdlMDVmYjZhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Rp5KLWNHP0160-Hhf6W803iBmdDcgfyuJuAp1ZRj73U">
> 
> Rendering now: <img alt="image" width="847" height="284" src="https://private-user-images.githubusercontent.com/406931/528988439-c09406f3-71c7-4c68-b96b-1fd3b2609643.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc1NTU1MTYsIm5iZiI6MTc2NzU1NTIxNiwicGF0aCI6Ii80MDY5MzEvNTI4OTg4NDM5LWMwOTQwNmYzLTcxYzctNGM2OC1iOTZiLTFmZDNiMjYwOTY0My5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMTA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDEwNFQxOTMzMzZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lMWYwM2FiOGM0N2ViZmYwYWNlYjdhMTg4MzI3YTM3YzgyZjBiYjIyYzkwYmUxZmM4Y2U4NDQxODNlYTMwMjI4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.GBFCkU4bjbmxkmOB48Qns5r4QEQcmiaETrJ6Pd8YmzI">
> 
> The url is not clickable and the line breaks are missing.
> 
> I noticed that it persists in html rather than markdown. I think Markdown is a better choice here
> 
> * It is a much simpler format. We only needs links and basic formatting. Html is overly complex for our needs.
> * Markdown is a simple, human readable format. Persisted content where the user has not added any formatting is cluttered with p tags.
> * Markdown is backwards compatible with plaintext. Meaning, the plaintext content we already have can be rendered as markdown. Especially if users used links, they can be rendered, as can basic formatting like underscores and asterisks.
> * Not a trap door. A markdown renderer can be removed and the content still looks good. HTML would require a migration.
> * Security. I'm sure tiptap renderer has xss mitigations, but it still makes me nervous about what happens if a story is submitted (especially bypassing the editor) with script tags, images, etc. I feel more confident in React's built-in escapes that are currently used on the plaintext content. fwiw I did try to put it script and img tags and the tiptap renderer stripped them out, and the header rendered as a p.
> 
> I see tiptap does support markdown to some extent. I think it's also worth looking at `react-markdown` for rendering. If you enable the gfm (Github flavored markdown) plugin in react-markdown it appears to enable auto-linking. It also support allow-listing tags.

Thanks for finding a good story for testing compatibility, I was having a hard time finding something more complex to see how tiptap would handle it. I mostly reached for Tiptap because it's a library I have experience using :D I'll take a look at react-markdown, thanks for the thorough feedback!

garcia5 committed:
chore: add react-markdown dependency

garcia5 committed:
feat(stories): render markdown for story text

- use `Markdown` component from react-markdown to convert markdown text -> HTML nodes
- limit to only basic elements, no headings, strikethroughs, etc. allowed
- add logic to automatically transform plain URLs into markdown links when story is rendered
- add custom <a> tag implementation for react-markdown to make sure links always open in new tabs

garcia5 commented:
Sorry for the delay here @jboolean!

Reworked this to be a lot more simple and use just `react-markdown`. Looks to be working with the example story you mentioned and I added in some basic auto linking functionality without including the gfm package.

<img width="650" height="453" alt="image" src="https://github.com/user-attachments/assets/d3c1a9da-740b-4c10-9a61-8292a5e314ba" />

I'm kind of torn on whether or not I should add to the `TextContent` placeholder text to indicate that markdown is supported, any feelings one way or another? I think it's probably better to leave the placeholder as-is to not confuse people who don't know what markdown is, but also could encourage nicer story formatting if people know it's supported.

Any thoughts?
